controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
-	if err := r.ReconcileNormal(rsCtx); err != nil {
+	result, err := r.ReconcileNormal(rsCtx)
+	if err != nil {
 		rsCtx.Logger.Error(err, "Failed to reconcile VirtualMachineReplicaSet")
 		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{}, nil
+	return result, nil
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet Deletion")
+
+	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
+		defer func() {
+			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", nil, false)
+		}()
+
+		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	}
+
+	return nil
+}
+
+// shouldExcludeVirtualMachine returns true if the VirtualMachine should be
+// filtered out, false otherwise.
+func shouldExcludeVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet, vm *vmopv1.VirtualMachine) bool {
+	if metav1.GetControllerOf(vm) != nil && !metav1.IsControlledBy(vm, rs) {
+		return true
+	}
+
+	return false
 }
 
-func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+// adoptOrphan sets the VirtualMachineReplicaSet as a controller OwnerReference
+// to the VirtualMachine.
+func (r *Reconciler) adoptOrphan(
+	ctx *pkgctx.VirtualMachineReplicaSetContext,
+	rs *vmopv1.VirtualMachineReplicaSet,
+	vm *vmopv1.VirtualMachine) error {
+
+	_, err := controllerutil.CreateOrPatch(ctx, r.Client, vm, func() error {
+		return controllerutil.SetControllerReference(rs, vm, r.Scheme())
+	})
+
+	return err
+}
+
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) (ctrl.Result, error) {
 	if !controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		// Set the finalizer and return immediately so the object is patched immediately.
+		// Set the finalizer and return so the object is patched immediately.
 		controllerutil.AddFinalizer(ctx.ReplicaSet, finalizerName)
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet")
 
-	return nil
+	if ctx.ReplicaSet.Labels == nil {
+		ctx.ReplicaSet.Labels = make(map[string]string)
+	}
+
+	if ctx.ReplicaSet.Spec.Selector.MatchLabels == nil {
+		ctx.ReplicaSet.Spec.Selector.MatchLabels = make(map[string]string)
+	}
+
+	if ctx.ReplicaSet.Spec.Template.Labels == nil {
+		ctx.ReplicaSet.Spec.Template.Labels = make(map[string]string)
+	}
+
+	selectorMap, err := metav1.LabelSelectorAsMap(ctx.ReplicaSet.Spec.Selector)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err,
```

Comment:
We are not using `github.com/pkg/errors` anymore. Please change this to:

```go
fmt.Errorf("failed to convert VirtualMachineReplicaSet %q label selector to a map: %w", ctx.ReplicaSet.Name, err)
```

Please see https://github.com/vmware-tanzu/vm-operator/pull/538 for more information.

Reply:
I was aware of your change.  Converted a few errors from "errors" -> "fmt" but missed a bunch :(  Thanks for pointing this.  Updating.